File: Problems/Algorithms/1631. Path with Minimum Effort/PathMinEffort.kt
class Solution {
    fun minimumEffortPath(heights: Array<IntArray>): Int {
        val n = heights.size
        val m = heights[0].size
        
        val deltas = arrayOf(intArrayOf(0, 1), intArrayOf(0, -1), intArrayOf(1, 0), intArrayOf(-1, 0))
        
        val visited = Array(n) { BooleanArray(m) { false } }
        val dp = Array(n) { IntArray(m) { Int.MAX_VALUE } }
        val heap = PriorityQueue<IntArray>({ a, b -> a[2] - b[2] })
        
        heap.add(intArrayOf(0, 0, 0))
        dp[0][0] = 0
        while (!heap.isEmpty()) {
            val curr = heap.poll()
            val row = curr[0]
            val col = curr[1]
            val effort = curr[2]
            
            if (!visited[row][col]) {
                visited[row][col] = true
                
                for (neighbor in deltas) {
                    val newRow = row + neighbor[0]
                    val newCol = col + neighbor[1]
                    
                    if (isValid(n, m, newRow, newCol, visited)) {
                        val newEffort = maxOf(effort, Math.abs(heights[newRow][newCol]-heights[row][col]))
                        
                        if (newEffort < dp[newRow][newCol]) {
                            dp[newRow][newCol] = newEffort
                            heap.add(intArrayOf(newRow, newCol, newEffort))
                        }
                    }
                }
            }
        }
        
        return dp[n-1][m-1]
    }
    
    private fun isValid(n: Int, m: Int, i: Int, j: Int, visited: Array<BooleanArray>): Boolean {
        return i >= 0 && i < n && j >= 0 && j < m && !visited[i][j]
    }
}

File: Problems/Algorithms/1631. Path with Minimum Effort/path_min_effort.py
#!/usr/bin/env python3

import math
from typing import List
from heapq import heappop, heappush


class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        rows, cols = len(heights), len(heights[0])
        distances = [[math.inf for _ in range(cols)] for _ in range(rows)]
        distances[0][0] = 0

        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        min_heap = [(0, 0, 0)]
        while min_heap:
            x, y, effort = heappop(min_heap)

            if x == rows - 1 and y == cols - 1:
                return effort

            for dx, dy in directions:
                nx, ny = x + dx, y + dy

                if 0 <= nx < rows and 0 <= ny < cols:
                    new_effort = max(effort, abs(heights[x][y] - heights[nx][ny]))

                    if new_effort < distances[nx][ny]:
                        distances[nx][ny] = new_effort
                        heappush(min_heap, (nx, ny, new_effort))
